lib/Dialect/Ibis/Transforms/IbisContainerize.cpp

```diff
@@ -0,0 +1,134 @@
+//===- IbisContainerize.cpp - Implementation of containerizing ------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+
+#include "circt/Dialect/Ibis/IbisDialect.h"
+#include "circt/Dialect/Ibis/IbisOps.h"
+#include "circt/Dialect/Ibis/IbisPasses.h"
+#include "circt/Dialect/Ibis/IbisTypes.h"
+
+#include "circt/Support/SymCache.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace circt;
+using namespace ibis;
+
+// Iterates the symbolcache until a unique name is found.
+static StringAttr getUniqueName(mlir::MLIRContext *ctx, StringAttr baseName,
+                                SymbolCache &symCache) {
+  StringAttr uniqueName = baseName;
+  int uniqueCntr = 0;
+  while (symCache.getDefinition(uniqueName))
+    uniqueName =
+        StringAttr::get(ctx, baseName.strref() + "_" + Twine(uniqueCntr++));
+  return uniqueName;
+}
+namespace {
+
+struct OutlineContainerPattern : public OpConversionPattern<ContainerOp> {
+  OutlineContainerPattern(MLIRContext *context, SymbolCache *symCache)
+      : OpConversionPattern<ContainerOp>(context), symCache(symCache) {}
+
+  using OpAdaptor = typename OpConversionPattern<ContainerOp>::OpAdaptor;
+
+  LogicalResult
+  matchAndRewrite(ContainerOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Outline the container into the module scope, by prefixing it with the
+    // parent class name.
+    auto parentClass =
+        dyn_cast_or_null<ClassOp>(op.getOperation()->getParentOp());
+    if (!parentClass)
+      return failure();
+
+    rewriter.setInsertionPoint(parentClass);
+    auto newContainerName =
+        rewriter.getStringAttr(parentClass.getName() + "_" + op.getName());
+    // unique it...
+    newContainerName =
+        getUniqueName(rewriter.getContext(), newContainerName, *symCache);
+    auto newContainer =
+        rewriter.create<ContainerOp>(op.getLoc(), newContainerName);
+    symCache->addDefinition(newContainerName, newContainer);
+
+    rewriter.mergeBlocks(op.getBodyBlock(), newContainer.getBodyBlock(), {});
+    rewriter.eraseOp(op);
+
+    // Rename the ibis.this operation to refer to the proper op.
+    auto thisOp =
+        cast<ThisOp>(cast<ScopeOpInterface>(*newContainer.getOperation())
+                         .getThis()
+                         .getDefiningOp());
+    rewriter.setInsertionPoint(thisOp);
+    rewriter.replaceOpWithNewOp<ThisOp>(thisOp, newContainer.getNameAttr());
+    return success();
```

Comment:
Shouldn't this leave behind something in the class to indicate that "this container which got moved needs to be instantiated here". Or something in the container op to indicate which class is the parent it got moved out of?

Reply:
> Or something in the container op to indicate which class is the parent it got moved out of?

Any meaningful use of the parent within the container has to come from a reference to the parent class.
I.e. for
```mlir
ibis.class @C {
  %this = ibis.this @C
  ibis.port.input @C_in : i1
  ibis.port.output @C_out : i1

  ibis.container @D {
    %this_d = ibis.this @D
    %parent_C = ibis.get_parent_of_ref %this_d : !ibis.scoperef<@D> -> !ibis.scoperef<@C>
    // Test local read/writes
    %c_in_p = ibis.get_port %parent_C, @C_in : !ibis.scoperef<@C> -> !ibis.portref<i1>
    %c_out_p = ibis.get_port %parent_C, @C_out : !ibis.scoperef<@C> -> !ibis.portref<i1>
    %true = hw.constant true
    ibis.port.write %c_out_p, %true : i1
    %c_out = ibis.port.read %c_in_p : !ibis.portref<i1>
  }
}
```
all accesses to the parent class are performed through the `get_parent_of_ref` operation. In the same vein, there are only accesses from within the container to the parent class, and not the other direction. So apart from renaming the container, there isn't (as far as i can tell) anything else that should be modified.

Reply:
Woops, correct. Added `ContainerInstanceOp`, lowering from the class `InstanceOp` to said op + addition of an instance op during container outlining.